core/vm/program.go

```diff
@@ -415,98 +386,163 @@ func jitCalculateGasAndSize(env Environment, contract *Contract, instr instructi
-		gas.Add(gas, words.Mul(words, params.CopyGas))
+		words := toWordSize(stack.data[stack.len()-3].Uint64())
+		if !math.IsMulSafe(words, KeccakWordGas64) {
+			return 0, 0, OutOfGasError
+		}
+		wordsGas := words * CopyGas64
+		if !math.IsAddSafe(gas, wordsGas) {
+			return 0, 0, OutOfGasError
+		}
+		gas += wordsGas
 	case CODECOPY:
-		newMemSize = calcMemSize(stack.peek(), stack.data[stack.len()-3])
+		newMemSize, sizeFault = calcMemSize(stack.peek(), stack.data[stack.len()-3])
+		if sizeFault {
+			return 0, 0, OutOfGasError
+		}
 
-		words := toWordSize(stack.data[stack.len()-3])
-		gas.Add(gas, words.Mul(words, params.CopyGas))
+		words := toWordSize(stack.data[stack.len()-3].Uint64())
+		if !math.IsMulSafe(words, KeccakWordGas64) {
+			return 0, 0, OutOfGasError
+		}
+		wordsGas := words * CopyGas64
+		if !math.IsAddSafe(gas, wordsGas) {
+			return 0, 0, OutOfGasError
+		}
+		gas += wordsGas
 	case EXTCODECOPY:
-		newMemSize = calcMemSize(stack.data[stack.len()-2], stack.data[stack.len()-4])
-
-		words := toWordSize(stack.data[stack.len()-4])
-		gas.Add(gas, words.Mul(words, params.CopyGas))
+		newMemSize, sizeFault = calcMemSize(stack.data[stack.len()-2], stack.data[stack.len()-4])
+		if sizeFault {
+			return 0, 0, OutOfGasError
+		}
 
+		words := toWordSize(stack.data[stack.len()-4].Uint64())
+		if !math.IsMulSafe(words, KeccakWordGas64) {
+			return 0, 0, OutOfGasError
+		}
+		wordsGas := words * CopyGas64
+		if !math.IsAddSafe(gas, wordsGas) {
+			return 0, 0, OutOfGasError
+		}
+		gas += wordsGas
 	case CREATE:
-		newMemSize = calcMemSize(stack.data[stack.len()-2], stack.data[stack.len()-3])
+		newMemSize, sizeFault = calcMemSize(stack.data[stack.len()-2], stack.data[stack.len()-3])
 	case CALL, CALLCODE:
-		gas.Add(gas, stack.data[stack.len()-1])
+		callGas := stack.data[stack.len()-1]
+		if callGas.BitLen() > 64 {
+			return 0, 0, OutOfGasError
+		}
+		gas += callGas.Uint64()
 
 		if op == CALL {
 			if !env.Db().Exist(common.BigToAddress(stack.data[stack.len()-2])) {
-				gas.Add(gas, params.CallNewAccountGas)
+				if !math.IsAddSafe(gas, CallNewAccountGas64) {
+					return 0, 0, OutOfGasError
+				}
+				gas += CallNewAccountGas64
 			}
 		}
 
 		if len(stack.data[stack.len()-3].Bytes()) > 0 {
-			gas.Add(gas, params.CallValueTransferGas)
+			if !math.IsAddSafe(gas, CallValueTransferGas64) {
+				return 0, 0, OutOfGasError
+			}
+			gas += CallValueTransferGas64
 		}
 
-		x := calcMemSize(stack.data[stack.len()-6], stack.data[stack.len()-7])
-		y := calcMemSize(stack.data[stack.len()-4], stack.data[stack.len()-5])
+		x, xSizeFault := calcMemSize(stack.data[stack.len()-6], stack.data[stack.len()-7])
+		if xSizeFault {
+			return 0, 0, OutOfGasError
+		}
+		y, ySizeFault := calcMemSize(stack.data[stack.len()-4], stack.data[stack.len()-5])
+		if ySizeFault {
+			return 0, 0, OutOfGasError
+		}
 
-		newMemSize = common.BigMax(x, y)
+		newMemSize = uint64(gmath.Max(float64(x), float64(y)))
```

Comment:
I think what he means is that float64 should not be used here. It is easy to avoid by doing

```go
newMemSize = x
if y > newMemSize {
    newMemSize = y
}
```